File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
public:
    int binarySearch(vector<int>& nums, int target, int s, int e){
        while (s <= e){
            int m = (s+e)/2;

            if (nums[m] == target) return m;
            if (nums[m] < target){
                s = m+1;
            }else{
                e = m-1;
            }
        }

        return -1;
    }

    int search(vector<int>& nums, int target) {
        if (nums.size() == 1){
            if (target == nums[0]) return 0;
            return -1;
        }

        if (nums[0] < nums[nums.size() - 1]){
            // not rotated
            return binarySearch(nums, target, 0, nums.size()-1);
        }

        // find switch point
        int init = nums[0];
        int s = 0;
        int e = nums.size() -1;
        int m = 0;

        while (s <= e){
            m = (s+e)/2;

            if (nums[m] > nums[m+1]){
                // Found cutoff
                break;
            }

            if (nums[m] >= init){
                // Part of list 1
                s = m+1;
            }else{
                // Part of list 2
                e = m-1;
            }

        }

        if  (target >= init && target <= nums[m]){
            return binarySearch(nums, target, 0, m);

        }else if (target >= nums[m+1] && target <= nums[nums.size()-1]){
            return binarySearch(nums, target, m+1, nums.size()-1);
        }

        return -1;
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    # Classic binary search question, do it in two phases
    # First we look for the 'drop', ie first val < nums[0] is the real start.
    # Then we check [0], min, [-1] to know which half the value is in, and do regular BS

    def searchForDrop(self, nums):    
        low, high = 0, len(nums) - 1
        best = -1

        while low <= high:
            mid = (low + high) // 2
            if nums[mid] < nums[0]:
                best = mid
                high = mid - 1
            else:
                low = mid + 1

        return best
    
    def regularBS(self, nums, target, low, high):
        while low <= high:
            mid = (low + high) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                high = mid - 1
            else:
                low = mid + 1

        return -1


    def search(self, nums: List[int], target: int) -> int:
        low, high = 0, len(nums) - 1
        if nums[0] > nums[-1]:
            realMin = self.searchForDrop(nums)
            if target >= nums[0]:
                high = realMin - 1
            else:
                low = realMin
        
        return self.regularBS(nums, target, low, high)